### PR TITLE
C51-62: Integrate functionality next milestone card and next activity card

### PR DESCRIPTION
**Pull request by swastikpareek**
## Overview 
This PR contains basic UI enhancements and integration of next activity and next milestone card

## Before 

c51-62 - before

## After
c51-62 - loading state
c51-62 - no result

c51-62- after the most recent

### Activity card states
 * With icon and type = task 
  c51-62 - act with icon and task

 * With icon 
   c51-62 - activity with icon

 * With task
  c51-62 - activity with task

 * Without icon and task
  c51-62 - activity without icon

## Technical Details

* Moved `isOverdue` and `formatDate` function into a new `DateHelper` service and added unit tests for them.
* Added three different variants for activity card 
  *  Activity Card Big - `ActivityCard--Big.html` and `_activity-card--big.scss`
  *  Activity Card Long - `ActivityCard--Long.html` and `_activity-card--long.scss`
  *  Activity Card Short - `ActivityCard--Short.html`  and `_activity-card--short.scss`
* Added two new attributes to activity card directive
 * `mode` - To switch between above mentioned visual modes - default to 'short'
 * `type` - To set the inner styles and no result behaviours of different types of activity the cards.

* Added new classes
  * `civicase__activity-card--big--milestone` : For milestone specific css

* Added empty state for actitivity card styles inside `_activity-no-result.scss` and new classes added
  * `civicase__no-result-icon` `civicase__no-result-icon--milestone` `civicase__activity-no-result-icon--activity`, `civicase__activity-no-result-icon--communication`, `civicase__activity-no-result-icon--tasks`.

* Update activity return params to fetch more required activity fields.

* Added tags related info in `GetDetails.php` file to get tags related info in activity cards.

* Changed `civicase__big-date` class convention to `civicase__highlighted-date` class
* Added hover styling to cards
* Fix case view tabs panel z-index and glicth issue.
* New classe `civicase__activity-card-left` has been added for taking care of left section of big activity card.

## Updates
 * Fix checkbox check issue for next activity being a task 
 * Add `star`  and `attachment` icons to the next activity and next mileston 
 * Theme the dropdown with the icons and styles
 * Update logic to each activity card which is milestone to have a red ribbon as a background .
c51-62 - after setting the label for milestone

### CODE REVIEW

**Comment by deb1990** [comment]
UI Issues
1. On hover, there is a blue border, is it in the design?
2018-09-21 at 11 09 am
we should follow the above for hover, unless specified

2. 3 dots design does not match with mock. Maybe we are using wrong icon and need to use from material.

3. When next activity is of task type, do we get the checkbox? please confirm

4. When there is no icon set for the next activity, how does it look?

5. Have we heard anything from Jon about the "Star" and "Attachment"?

File: ang/civicase/Utils.js [hunk 1]
```diff
@@ -623,6 +623,31 @@
...
+    };
+
+    /**
+     * Formats Date in correct format (DD/MM/YYYY)
+     *
+     * @param {String} date ISO string
+     * @param {String} format ISO string
+     * @return {String} the formatted date
+     */
+    this.formatDate = function (date, format = 'DD/MM/YYYY') {
```
Comment (deb1990) [inline ang/civicase/Utils.js]: `format = 'DD/MM/YYYY'` This is ES6 , which we are not using for now. Please use ES5.
```
this.formatDate = function (date, format) {
      format = format || 'DD/MM/YYYY';
      
      return moment(date).format(format);
    };
```

File: ang/test/civicase/Utils.spec.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
+/* eslint-env jasmine */
```
Comment (deb1990) [inline ang/test/civicase/Utils.spec.js]: According to the above comment, the file name should change.

File: ang/test/civicase/Utils.spec.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    beforeEach(module('civicase'));
+
+    describe('DateHelper', function () {
+      beforeEach(inject(function (_DateHelper_) {
+        DateHelper = _DateHelper_;
+      }));
+
+      describe('formatDate()', function () {
+        it('checks the format of the date', function () {
```
Comment (deb1990) [inline ang/test/civicase/Utils.spec.js]: Wrong description.
It does not check the format, it returns the date in the sent format.

File: ang/test/civicase/Utils.spec.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        });
+      });
+
+      describe('isOverdue()', function () {
+        beforeEach(function () {
+          pastDate = moment().subtract(7, 'd');
+          futureDate = moment().add(7, 'd');
+        });
+
+        it('checks the date is overdue ', function () {
```
Comment (deb1990) [inline ang/test/civicase/Utils.spec.js]: `returns true when date is in past`

File: ang/test/civicase/Utils.spec.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+          expect(DateHelper.isOverdue(pastDate)).toBe(true);
+        });
+
+        it('checks the date is not overdue ', function () {
```
Comment (deb1990) [inline ang/test/civicase/Utils.spec.js]: `returns false when date is not in past`

File: icons/activities.svg [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (deb1990) [inline icons/activities.svg]: Now that we have different kinds of resources. Its better to collect them under a folder.
So please create a folder `resource`. And move `fonts` and `icons` inside.

File: scss/components/_big-date.scss [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /** Big calendar type date styling **/
-.civicase__big-date {
+.civicase__highlighted-date {
```
Comment (deb1990) [inline scss/components/_big-date.scss]: File name and comments should change to follow this changed component name.

File: scss/components/_activity-no-result-icon.scss [hunk 8]
```diff
@@ -0,0 +1,15 @@
+.civicase__no-result-icon {
```
Comment (deb1990) [inline scss/components/_activity-no-result-icon.scss]: These icons are only for activity right?
The component name does not denote that.
Should be 
`civicase__activity-no-result-icon` - this one should have the SVG for Activities which is not milestone.
`civicase__activity-no-result-icon--milestone`
`civicase__activity-no-result-icon--task` (no need to create now, but for future)
`civicase__activity-no-result-icon--communication` (no need to create now, but for future)

Lastly, this is not a new component, please place them under `ActivityCard.scss`

File: scss/components/_activity-card.scss [hunk 9]
```diff
@@ -111,9 +110,6 @@
 .civicase__activity-subject {
   color: $gray-darker;
   font-weight: $font-weight-heading;
-  overflow: hidden;
```
Comment (deb1990) [inline scss/components/_activity-card.scss]: Why are these removed?

File: ang/civicase/CaseDetails.js [hunk 10]
```diff
@@ -334,9 +334,34 @@
...
+     * @return {object} - next activity
+     */
+    function findNextActivityWhichIsNotMilestone (activities) {
+      var nextActivity;
+      var hit = false;
+
+      _.each(activities, function (act, key) {
+        if (CRM.civicase.activityTypes[act.activity_type_id].grouping !== 'milestone' && !hit) {
+          nextActivity = act;
+          nextActivity.type = CRM.civicase.activityTypes[act.activity_type_id].label;
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply ang/civicase/CaseDetails.js]: `Next Milestone` and `next_activity` are pre-populated with these keys from `case/GetDetails.php` file. If we wish to do it for `activities` coming from another Backend API route. We will need to do it at the backend and will require more efforts and time. And also as per our discussion (which we follow for civiHR development) its better to manipulate data at the front end.

Please let me know if you have any doubts.

File: scss/components/_activity-no-result-icon.scss (hunk 8 above)
Comment (swastikpareek) [reply scss/components/_activity-no-result-icon.scss]: > Should be `civicase__activity-no-result-icon` - Noted 
>  this is not a new component, please place them under `ActivityCard.scss` - Noted

> this one should have the SVG for Activities which is not milestone. 
I think according to BEM. There should be a base class which contains all the basic generic styles
and then these styles should have different variants.
so it should be something like this
```
.civicase__activity-no-result-icon {
  background-position: center center;
  background-repeat: no-repeat;
  background-size: contain;
  height: 48px;
  width: 48px;
}

.civicase__activity-no-result-icon--milestone {
  background-image: url('../resources/icons/milestone.svg');
}

.civicase__activity-no-result-icon--activity {
  background-image: url('../resources/icons/activities.svg');
}

.civicase__activity-no-result-icon--activity {
  background-image: url('../resources/icons/activities.svg');
}

.civicase__activity-no-result-icon--activity {
  background-image: url('../resources/icons/activities.svg');
}

```

File: scss/components/_activity-card.scss (hunk 9 above)
Comment (swastikpareek) [reply scss/components/_activity-card.scss]: These are not removed. These are moved under `_activity-card--short.scss` as we need these styles for short activity card variant only.

**Comment by swastikpareek** [comment]
@deb1990 
 
> 1. On hover, there is a blue border, is it in the design?
>    2018-09-21 at 11 09 am
>    we should follow the above for hover, unless specified
> 2. 3 dots design does not match with mock. Maybe we are using wrong icon and need to use from material.
> 3. When next activity is of task type, do we get the checkbox? please confirm
> 4. When there is no icon set for the next activity, how does it look?
> 5. Have we heard anything from Jon about the "Star" and "Attachment"?

1. Hover effect is updated
2. Dot icon now being used from material design. Please refer screenshot.
3. Please find the attached screenshots for it.
4. Same as 3.
5. Yes. Jon has over-sighted this he said. It is good to have these on the card but we need to get this clarified from @shitijg - he added.
 As per our discussion we can have this discussed later w/o blocking this PR.

File: ang/civicase/ActivityCard.js [hunk 11]
```diff
@@ -4,21 +4,33 @@
   module.directive('caseActivityCard', function () {
     return {
       restrict: 'A',
-      templateUrl: '~/civicase/ActivityCard.html',
+      templateUrl: function (elem, attrs) {
+        switch (attrs.mode) {
+          case 'big':
+            return `~/civicase/ActivityCard--Big.html`;
```
Comment (deb1990) [inline ang/civicase/ActivityCard.js]: You have used ` instead of `'`

File: ang/civicase/ActivityCard.js [hunk 12]
```diff
@@ -4,21 +4,33 @@
...
     };
   });
 
-  function caseActivityCardController ($scope, getActivityFeedUrl, dialogService, templateExists, crmApi) {
+  function caseActivityCardController ($scope, getActivityFeedUrl, dialogService, templateExists, crmApi, DateHelper) {
     var ts = $scope.ts = CRM.ts('civicase');
     $scope.activityFeedUrl = getActivityFeedUrl;
     $scope.formatDate = CRM.utils.formatDate;
     $scope.templateExists = templateExists;
+    $scope.formatDate = DateHelper.formatDate;
```
Comment (deb1990) [inline ang/civicase/ActivityCard.js]: `$scope.formatDate` function is defined twice. Also on line 30

File: ang/civicase/CaseListTable.js [hunk 13]
```diff
@@ -241,30 +241,34 @@
...
+          sort: 'activity_date_time ASC'
+        },
+        return: [
+          'subject', 'details', 'activity_type_id', 'status_id', 'source_contact_name',
+          'target_contact_name', 'assignee_contact_name', 'activity_date_time', 'is_star',
+          'original_id', 'tag_id.name', 'tag_id.description', 'tag_id.color', 'file_id',
+          'is_overdue', 'case_id'
+        ]
+      };
       var returnParams = {
```
Comment (deb1990) [inline ang/civicase/CaseListTable.js]: This should be renamed to `returnCaseParams` to avoid confusion.

File: ang/civicase/DateHelperService.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+     *
+     * @param {String} date ISO string
+     * @return {Boolean} if the date is overdue.
+     */
+    this.isOverdue = function (date) {
+      return moment(date).isBefore(moment());
+    };
+
+    /**
+     * Formats Date in correct format (DD/MM/YYYY)
```
Comment (deb1990) [inline ang/civicase/DateHelperService.js]:
```
* Formats Date in sent format
* Default format is (DD/MM/YYYY)
```

File: ang/test/civicase/DateHelperService.spec.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
+/* eslint-env jasmine */
+(function ($) {
+  describe('Utils', function () {
```
Comment (deb1990) [inline ang/test/civicase/DateHelperService.spec.js]: Utils should be removed as its not part of Utils anymore.

File: scss/components/_activity-card.scss [hunk 16]
```diff
@@ -216,3 +216,27 @@
...
+.civicase__activity-no-result-icon--milestone {
+  background-image: url('../resources/icons/milestone.svg');
+}
+
+.civicase__activity-no-result-icon--activity {
+  background-image: url('../resources/icons/activities.svg');
+}
+
+.civicase__activity-no-result-icon--communication {
+  background-image: url('../resources/icons/activities.svg');
```
Comment (deb1990) [inline scss/components/_activity-card.scss]: Wrong svg?

File: scss/components/_activity-card.scss [hunk 17]
```diff
@@ -216,3 +216,27 @@
...
+}
+
+.civicase__activity-no-result-icon--tasks {
+  background-image: url('../resources/icons/activities.svg');
```
Comment (deb1990) [inline scss/components/_activity-card.scss]: Wrong svg?

**Comment by deb1990** [comment]
@swastikpareek When you Delete Next Activity or Milestone, does the UI refresh to show latest data?

**Comment by deb1990** [comment]
@swastikpareek There is no hover effect on the 3 dot icons. I was there previously.

**Comment by swastikpareek** [comment]
> @swastikpareek There is no hover effect on the 3 dot icons. I was there previously. 

Updated the styles. Please test

> @swastikpareek When you Delete Next Activity or Milestone, does the UI refresh to show latest data?

Fixed this as well
